### PR TITLE
Closes #20345 Adds minhash filter for use in Elasticsearch

### LINKED ISSUE #20345: Add minhash token filter
It would be nice to be able to use the minhash token filter, using [minHash token filter](https://issues.apache.org/jira/browse/LUCENE-6968) added to Lucene.

The minhash algorithm implements locality-sensitive hashing. Basically, every token is hashed into a hash table. The hash table is then sampled to generate a smaller set of representative tokens.

This has several applications
- Dimensionality reduction that preserves information that can be used in similarity calculations. For example, efficient deduplication over large sets of text, more like this, etc.

### PR DESCRIPTION
**Pull request by softwaredoug**
- Adds minhash token filter, using [minHash token filter](https://issues.apache.org/jira/browse/LUCENE-6968) added to Lucene.

The minhash algorithm implements locality-sensitive hashing. Basically, every token is hashed into a hash table. The hash table is then sampled to generate a smaller set of representative tokens.

This has several applications
- Dimensionality reduction that preserves information that can be used in similarity calculations. For example, efficient deduplication over large sets of text, more like this, etc.

The minhash filter has 4 parameters
- bucket_size (def 512) how many buckets to use in the hash table. Will also be the number of tokens output from the minhash filter. More buckets means slower performance and more space, but greater precision.
- num_hash_tables (def 1). As hash tables are sampled by looking at the first item in the bucket, adding more hash tables will provide added precision
- bucket_depth (def 1). How many items should be stored in each bucket?
- rotation (def true if num_buckets > 1). For any empty buckets, fills them with the contents of the next bucket

Implementation Notes/Questions
- Are tests needed above the Lucene tests that already exist?
- Do I need to provide documentation as well?
- You'll note the fwd'ing to the Lucene filter factory. This is because of an annoying issue with the MinHashFilter. See this [Lucene issue](https://issues.apache.org/jira/browse/LUCENE-7436)

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/analysis/MinHashTokenFilterFactory.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import org.apache.lucene.analysis.minhash.MinHashFilterFactory;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.index.IndexSettings;
+
+import java.util.HashMap;
+import java.util.Map;
+
+
+public class MinHashTokenFilterFactory extends AbstractTokenFilterFactory {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/analysis/MinHashTokenFilterFactory.java]: can this be final?

File: core/src/main/java/org/elasticsearch/index/analysis/MinHashTokenFilterFactory.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    private MinHashFilterFactory luceneFactory;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/analysis/MinHashTokenFilterFactory.java]: this should also be final?

File: core/src/main/java/org/elasticsearch/index/analysis/MinHashTokenFilterFactory.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+     * Create a {@link org.apache.lucene.analysis.minhash.MinHashFilterFactory}.
+     */
+    public MinHashTokenFilterFactory(IndexSettings indexSettings, Environment environment, String name, Settings settings) {
+        super(indexSettings, name, settings);
+
+        // annoying settings indirection because MinHashFilter
+        // has package-private constructor and defaults :(
+        // see https://issues.apache.org/jira/browse/LUCENE-7436
+        Map<String, String> settingsMap = new HashMap<String, String>();
+        settingsMap.put("hashCount" ,settings.get("num_hash_tables"));
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/analysis/MinHashTokenFilterFactory.java]: I wonder if we should at least use the same terms just `_` separated?

File: core/src/main/java/org/elasticsearch/indices/analysis/AnalysisModule.java [hunk 4]
```diff
@@ -19,126 +19,14 @@
...
 package org.elasticsearch.indices.analysis;
 
+import org.apache.lucene.analysis.minhash.MinHashFilterFactory;
 import org.elasticsearch.Version;
 import org.elasticsearch.cluster.metadata.IndexMetaData;
 import org.elasticsearch.common.NamedRegistry;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.env.Environment;
 import org.elasticsearch.index.IndexSettings;
-import org.elasticsearch.index.analysis.ASCIIFoldingTokenFilterFactory;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/indices/analysis/AnalysisModule.java]: these changes are unrelated?

**Comment by s1monw** [comment]
Thanks for doing this!! I left some comments, I also wonder if we can get a simple test for all the parameters etc. like `org.elasticsearch.index.analysis.KeepFilterFactoryTests`. It also seems that you haven't signed the CLA, maybe you used a different mail address than in your git commits?

**Comment by s1monw** [comment]
> Do I need to provide documentation as well?

that would be fantastic

**Comment by jasontedor** [comment]
Isn't there already an open PR for this, #20206?

**Comment by s1monw** [comment]
> Isn't there already an open PR for this, #20206?

I agree the one you linked looks more complete, I think we should close this one?

**Comment by softwaredoug** [comment]
@jasontedor so there is! I swear I searched for minhash, now I see it. Sorry about the dup. @s1monw not sure if you have a procedure for closing dups?

@s1monw I'll look into the CLA for future PRs.

**Comment by s1monw** [comment]
thanks anyway @softwaredoug  ;) doesn't happen too often - keep it coming!
